File: Java/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        if(nums.length == k)
            return nums;
        Map<Integer, Integer> map = new HashMap();
        for(int i: nums){
            map.put(i, map.getOrDefault(i, 0) + 1);
        }

        Queue<Integer> head = new PriorityQueue<>(
            (n1, n2) -> map.get(n1) - map.get(n2));

        for(int i : map.keySet()){
            head.add(i);
            if(head.size() > k)
                head.poll();
        }
        int [] res = new int[k];
        for(int i = k -1; i >= 0; i--){
            res[i] = head.poll();
        }
        return res;
    }
}

File: C++/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<int> res;
        if( nums.size() == k)
            return nums;

    unordered_map<int, int> count_map;
    
    for(int i : nums){
        count_map[i]++;
    }
    priority_queue<pair<int, int>> pd;

    for(auto it = count_map.begin(); it != count_map.end(); it++){
        pd.push(make_pair(it->second, it->first));
        if(pd.size() > (int) count_map.size() - k){
            res.push_back(pd.top().second);
            pd.pop();
        }
    }
    return res;
    }
};
